hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java

```diff
@@ -1618,24 +1621,33 @@ private void setupPipelineForAppendOrRecovery() throws IOException {
       LOG.warn(msg);
       lastException.set(new IOException(msg));
       streamerClosed = true;
-      return;
+      return false;
     }
-    setupPipelineInternal(nodes, storageTypes, storageIDs);
+    return setupPipelineInternal(nodes, storageTypes, storageIDs);
   }
 
-  protected void setupPipelineInternal(DatanodeInfo[] datanodes,
+  protected boolean setupPipelineInternal(DatanodeInfo[] datanodes,
       StorageType[] nodeStorageTypes, String[] nodeStorageIDs)
       throws IOException {
     boolean success = false;
     long newGS = 0L;
+    boolean isCreateStage = BlockConstructionStage.PIPELINE_SETUP_CREATE == stage;
     while (!success && !streamerClosed && dfsClient.clientRunning) {
       if (!handleRestartingDatanode()) {
-        return;
+        return false;
+      }
+
+      final boolean isRecovery = errorState.hasInternalError() && !isCreateStage;
+
+      // During create stage, if we remove a node (nodes.length - 1)
+      //  min replication should still be satisfied.
+      if (isCreateStage && !(dfsClient.dtpReplaceDatanodeOnFailureReplication > 0 &&
```

Comment:
Reason behind adding this check here:
We are already doing this check in catch block of `addDatanode2ExistingPipeline` method [here](https://github.com/apache/hadoop/blob/trunk/hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java#L1528-L1539).
But when `isAppend` flag is set to `false` and we are in `PIPELINE_SETUP_CREATE` phase, we exit early from `addDatanode2ExistingPipeline` method [here](https://github.com/apache/hadoop/blob/trunk/hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java#L1489-L1492)

Lets say the replication factor is 3 and we have set the config property `dfs.client.block.write.replace-datanode-on-failure.min-replication` to 3 and there is one bad node in the pipeline.  Even if we have set the config property to `ReplaceDatanodeOnFailure.CONDITION_TRUE`, the code will exit the addDatanode2ExistingPipeline method early since `isAppend` is set to false and stage is `PIPELINE_SETUP_CREATE`. Assuming that there are NO available nodes in the rack, the pipeline will succeed with 2 nodes in the pipeline which will violate the config property: `dfs.client.block.write.replace-datanode-on-failure.min-replication`

Having written all of these, I realized that even if there are some good nodes available in the rack, we will exit early after this patch. Should we move this check after `handleDatanodeReplacement` method?  @ritegarg

Reply:
Thinking about the case where we are in PIPELINE_SETUP_CREATE stage but isAppend is set to true, then it will not exit early from addDatanode2ExistingPipeline method [here](https://github.com/apache/hadoop/blob/trunk/hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java#L1489-L1492) 

Assuming the `replication factor is 3` and `dfs.client.block.write.replace-datanode-on-failure.min-replication is set to 3` and there is 1 bad node in the pipeline and there are valid nodes in the cluster, this patch will return false early. 
I think we should move this check after `handleDatanodeReplacement` method. @ritegarg